dnf/rpm/miscutils.py

```diff
@@ -13,47 +13,84 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 # Copyright 2003 Duke University
 
-from __future__ import print_function, absolute_import
-from __future__ import unicode_literals
+from __future__ import print_function, absolute_import, unicode_literals
 
-import rpm
 import os
 import subprocess
 import logging
-
-from dnf.i18n import ucd
-from dnf.i18n import _
 from shutil import which
 
+from dnf.i18n import _
 
-logger = logging.getLogger('dnf')
+_logger = logging.getLogger('dnf')
+_rpmkeys_binary = None
 
+def _find_rpmkeys_binary():
+    global _rpmkeys_binary
+    if _rpmkeys_binary is None:
+        _rpmkeys_binary = which("rpmkeys")
+        _logger.debug(_('Using rpmkeys executable at %s to verify signatures'),
+                      _rpmkeys_binary)
+    return _rpmkeys_binary
 
-def _verifyPkgUsingRpmkeys(package, installroot, fdno):
-    os.lseek(fdno, 0, os.SEEK_SET)
-    rpmkeys_binary = '/usr/bin/rpmkeys'
-    if not os.path.isfile(rpmkeys_binary):
-        rpmkeys_binary = which("rpmkeys")
-        logger.info(_('Using rpmkeys executable from {path} to verify signature for package: {package}.').format(
-            path=rpmkeys_binary, package=package))
+def _process_rpm_output(data):
+    # No signatures or digests = corrupt package.
+    # There is at least one line for -: and another (empty) entry after the
+    # last newline.
+    if len(data) < 3 or data[0] != b'-:' or data[-1]:
+        return 2
+    seen_sig, missing_key, not_trusted, not_signed = False, False, False, False
+    for i in data[1:-1]:
+        if b': BAD' in i:
+            return 2
+        elif i.endswith(b': NOKEY'):
+            missing_key = True
+        elif i.endswith(b': NOTTRUSTED'):
+            not_trusted = True
+        elif i.endswith(b': NOTFOUND'):
+            not_signed = True
+        elif not i.endswith(b': OK'):
+            return 2
+    if not_trusted:
+        return 3
+    elif missing_key:
+        return 1
+    elif not_signed:
+        return 4
+    # we still check return code, so this is safe
+    return 0
 
+def _verifyPackageUsingRpmkeys(package, installroot):
+    rpmkeys_binary = _find_rpmkeys_binary()
     if not os.path.isfile(rpmkeys_binary):
```

Comment:
Oh I am sorry for not noticing this before but when there is no `rpmkeys` binary this ends in an ugly traceback because `_find_rpmkeys_binary()` returns `None` and `isfile()` accepts only:
```
path: A path-like object representing a file system path. A path-like object is either a string or bytes object representing a path.
```
We have to check for that before calling `isfile()`.